Formula/espflash.rb

```diff
@@ -0,0 +1,19 @@
+class Espflash < Formula
+  desc "Serial flasher utility for Espressif SoCs and modules based on esptool.py"
+  homepage "https://github.com/esp-rs/espflash"
+  url "https://github.com/esp-rs/espflash/archive/refs/tags/v2.0.0.tar.gz"
+  sha256 "01b65bdd7c22b422d18db6cf186cb788151f0ca0eda9533ff9478aaf7255a2cf"
+  license any_of: ["MIT", "Apache-2.0"]
+
+  depends_on "rust" => :build
+  depends_on "libuv"
+
+  def install
+    system "cargo", "install", *std_cargo_args
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/espflash --version")
+    assert_match "A command-line tool for flashing Espressif devices", shell_output("#{bin}/espflash --help")
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
How about attempting to flash a non-existent external board, and checking to see if we get the expected exit code?

Reply:
that's okay, but just want to provide a more meaningful test, give me some time to test the output on with gh-action runner.

Reply:
can homebrew test capture an error output?

it seems when test command gives an error, whole test break

```
❯ brew test espflash
==> Testing espflash
==> /opt/homebrew/Cellar/espflash/2.0.0/bin/espflash --version
==> /opt/homebrew/Cellar/espflash/2.0.0/bin/espflash --help
==> /opt/homebrew/Cellar/espflash/2.0.0/bin/espflash flash esp32_hal_blinky --port COMX
Error: espflash::connection_failed

  x Error while connecting to device
  `-> Serial port not found
  help: Ensure that the device is connected and your host recognizes the
        serial adapter

Error: espflash: failed
An exception occurred within a child process:
  Minitest::Assertion: Expected /Ensure/ to match "".
/opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.18.1/lib/minitest/assertions.rb:183:in `assert'
/opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/minitest-5.18.1/lib/minitest/assertions.rb:295:in `assert_match'
/opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/espflash.rb:24:in `block (2 levels) in <class:Espflash>'
/opt/homebrew/Library/Homebrew/resource.rb:128:in `block (2 levels) in unpack'
/opt/homebrew/Library/Homebrew/download_strategy.rb:120:in `chdir'
/opt/homebrew/Library/Homebrew/download_strategy.rb:105:in `stage'
/opt/homebrew/Library/Homebrew/resource.rb:124:in `block in unpack'
/opt/homebrew/Library/Homebrew/mktemp.rb:75:in `block in run'
/opt/homebrew/Library/Homebrew/mktemp.rb:75:in `chdir'
/opt/homebrew/Library/Homebrew/mktemp.rb:75:in `run'
/opt/homebrew/Library/Homebrew/resource.rb:228:in `stage_resource'
/opt/homebrew/Library/Homebrew/resource.rb:123:in `unpack'
/opt/homebrew/Library/Homebrew/resource.rb:97:in `stage'
/opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/espflash.rb:23:in `block in <class:Espflash>'
/opt/homebrew/Library/Homebrew/formula.rb:2406:in `block (3 levels) in run_test'
/opt/homebrew/Library/Homebrew/extend/kernel.rb:499:in `with_env'
/opt/homebrew/Library/Homebrew/formula.rb:2405:in `block (2 levels) in run_test'
/opt/homebrew/Library/Homebrew/formula.rb:987:in `with_logging'
/opt/homebrew/Library/Homebrew/formula.rb:2404:in `block in run_test'
/opt/homebrew/Library/Homebrew/mktemp.rb:75:in `block in run'
/opt/homebrew/Library/Homebrew/mktemp.rb:75:in `chdir'
/opt/homebrew/Library/Homebrew/mktemp.rb:75:in `run'
/opt/homebrew/Library/Homebrew/formula.rb:2682:in `mktemp'
/opt/homebrew/Library/Homebrew/formula.rb:2398:in `run_test'
/opt/homebrew/Library/Homebrew/test.rb:44:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/opt/homebrew/Library/Homebrew/test.rb:48:in `<main>'
```

Reply:
I solve it by redirect the error output to a text file, now it's working, btw can anyone review these changes?